src/librustc/ty/flags.rs

```diff
@@ -248,7 +247,7 @@ impl FlagComputation {
                 self.add_flags(TypeFlags::HAS_FREE_LOCAL_NAMES | TypeFlags::HAS_PARAMS);
             }
             ConstValue::Placeholder(_) => {
-                self.add_flags(TypeFlags::HAS_FREE_REGIONS | TypeFlags::HAS_CT_PLACEHOLDER);
+                self.add_flags(TypeFlags::HAS_CT_PLACEHOLDER);
```

Comment:
I think you need `HAS_FREE_LOCAL_NAMES`?

Reply:
Despite not needing it for `ty::Placeholder`?
https://github.com/rust-lang/rust/blob/174ac114bdb2b6e719133f5b308ad8501d0a328a/src/librustc/ty/flags.rs#L116-L118

Reply:
Okay that's 3-way inconsistent though, either `Region` is wrong or `Ty` is. cc @nikomatsakis

Reply:
That's true, though I had assumed `Region` was special cased deliberately because of:
https://github.com/rust-lang/rust/blob/174ac114bdb2b6e719133f5b308ad8501d0a328a/src/librustc/infer/canonical/canonicalizer.rs#L513

Reply:
cc @scalexm who [added](https://github.com/varkor/rust/commit/91623ca6405b5e3c7736bcacb244718906d4aa69) `HAS_TY_PLACEHOLDER`.

Reply:
Hmm. So I think that "placeholder types" are effectively unused in rustc today, because we don't support things like `for<T> Foo: Bar<T>`. I don't quite understand what @eddyb is saying, in that region placeholders don't seem to have that flag either, but maybe I'm confused? 

In any case, the flag indicates content that is not "globally cacheable" -- I think we probably *should* set the flag for placeholders, since they reference universe indices, and those indices are local to a particular inference context, and don't have meaning across inference contexts in particular.

Reply:
`HAS_FREE_LOCAL_NAMES`  is `HAS_FREE_REGIONS` for regions, IIUC.

As for where placeholders are used today, I'm pretty sure canonical queries use them where inference variables are involved? But I may have gotten misled by some doc comments or confusing `Bound` with `Placeholder`.

Reply:
Okay, so we should probably have `HAS_FREE_LOCAL_NAMES` for all placeholders? I'll update the PR.

Reply:
Ah I missed that second `match`, that makes more sense indeed.